control-plane/controllers/resources/gateway_indices.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package resources
+
+import (
+	"context"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	meshv2beta1 "github.com/hashicorp/consul-k8s/control-plane/api/mesh/v2beta1"
+)
+
+const (
+	// Naming convention: TARGET_REFERENCE.
+	GatewayClass_GatewayClassConfigIndex = "__v2_gatewayclass_referencing_gatewayclassconfig"
+
+	Gateway_GatewayClassIndex = "__v2_gateway_referencing_gatewayclass"
+)
+
+// RegisterFieldIndexes registers all of the field indexes for the API gateway controllers.
+// These indexes are similar to indexes used in databases to speed up queries.
+// They allow us to quickly find objects based on a field value.
+func RegisterFieldIndexes(ctx context.Context, mgr ctrl.Manager) error {
+	for _, index := range indexes {
+		if err := mgr.GetFieldIndexer().IndexField(ctx, index.target, index.name, index.indexerFunc); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type index struct {
+	name        string
+	target      client.Object
+	indexerFunc client.IndexerFunc
+}
+
+var indexes = []index{
+	{
+		name:        GatewayClass_GatewayClassConfigIndex,
+		target:      &meshv2beta1.GatewayClass{},
+		indexerFunc: gatewayClassConfigForGatewayClass,
+	},
+	{
+		name:        Gateway_GatewayClassIndex,
+		target:      &gwv1beta1.Gateway{},
+		indexerFunc: gatewayClassForGateway,
+	},
+}
+
+// gatewayClassConfigForGatewayClass creates an index of every GatewayClassConfig referenced by a GatewayClass.
+func gatewayClassConfigForGatewayClass(o client.Object) []string {
+	gc := o.(*meshv2beta1.GatewayClass)
+
+	pr := gc.Spec.ParametersRef
+	if pr != nil && pr.Kind == "GatewayClassConfig" {
+		return []string{pr.Name}
+	}
+
+	return []string{}
+}
```

Comment:
I'd update the docstring and function names here to reflect that this function isn't actually doing any indexing itself and doesn't return the full type, just the name of one.

I'd also be fine with inlining it above since it doesn't appear to be re-used anywhere

```suggestion
// mapGatewayClassToGatewayClassConfigName returns the name of the GatewayClassConfig referenced by a GatewayClass.
func mapGatewayClassToGatewayClassConfigName(o client.Object) []string {
	gc := o.(*meshv2beta1.GatewayClass)

	pr := gc.Spec.ParametersRef
	if pr != nil && pr.Kind == "GatewayClassConfig" {
		return []string{pr.Name}
	}

	return []string{}
}
```